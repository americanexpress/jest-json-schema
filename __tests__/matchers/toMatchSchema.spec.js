/*
 * Copyright (c) 2017 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

const ajvKeywords = require('ajv-keywords');
const chalk = require('chalk');
const toMatchSchemaUnderTest = require('../..').matchers.toMatchSchema;
const toMatchSchemaWithFormatsUnderTest = require('../..').matchersWithFormats({
  bcp47: /^[a-z]{2}-[A-Z]{2}$/,
}).toMatchSchema;
const toMatchSchemaWithOptionsUnderTest = require('../..').matchersWithOptions({
  verbose: true,
}, (ajv) => {
  ajvKeywords(ajv, ['typeof', 'instanceof']);
}).toMatchSchema;

chalk.enabled = false;

expect.extend({
  toMatchSchemaUnderTest,
  toMatchSchemaWithFormatsUnderTest,
  toMatchSchemaWithOptionsUnderTest,
});

describe('toMatchSchema', () => {
  let schema;

  beforeEach(() => {
    schema = {
      properties: {
        hello: { type: 'string' },
      },
      required: ['hello'],
    };
  });

  it('does not throw', () => {
    expect({ hello: 'world' }).toMatchSchemaUnderTest(schema);
    expect({ hello: 'a', world: 'b' }).toMatchSchemaUnderTest(schema);
    expect({}).not.toMatchSchemaUnderTest(schema);
    expect({ hello: 1 }).not.toMatchSchemaUnderTest(schema);
  });

  it('fails for wrong type', () => {
    const testObj = { hello: 1 };
    expect(() => expect(testObj).toMatchSchemaUnderTest(schema))
      .toThrowErrorMatchingSnapshot();
  });

  it('fails for missing required keys', () => {
    expect(() => expect({}).toMatchSchemaUnderTest(schema))
      .toThrowErrorMatchingSnapshot();
  });

  it('fails when pattern does not match', () => {
    schema.properties.hello.pattern = '[a-z]+';
    const testObj = { hello: '123' };
    expect(() => expect(testObj).toMatchSchemaUnderTest(schema))
      .toThrowErrorMatchingSnapshot();
  });

  it('fails when additional properties are found but forbidden', () => {
    schema.additionalProperties = false;
    const testObj = {
      hello: 'world',
      another: 'property',
    };
    expect(() => expect(testObj).toMatchSchemaUnderTest(schema))
      .toThrowErrorMatchingSnapshot();
  });

  it('includes the description in the error when provided', () => {
    const testObj = { hello: 1 };
    expect(() => expect(testObj).toMatchSchemaUnderTest(schema, 'en-US language pack'))
      .toThrowErrorMatchingSnapshot();
  });

  it('fails for matching schema when using .not', () => {
    const testObj = { hello: 'world' };
    expect(() => expect(testObj).not.toMatchSchemaUnderTest(schema))
      .toThrowErrorMatchingSnapshot();
  });

  it('does not crash on circular references', () => {
    const testObj = {};
    testObj.hello = testObj;
    expect(() => expect(testObj).toMatchSchemaUnderTest(schema))
      .toThrowErrorMatchingSnapshot();
  });

  it('assertion error matcherResult property contains matcher name and actual value', () => {
    schema.additionalProperties = false;
    const testObj = { another: 'property' };
    try {
      expect(testObj).toMatchSchemaUnderTest(schema);
    } catch (error) {
      expect(error.matcherResult).toEqual({
        actual: testObj,
        message: expect.any(Function),
        name: 'toMatchSchema',
        pass: false,
      });
    }
  });

  describe('custom formats', () => {
    describe('bcp47', () => {
      beforeEach(() => {
        schema = {
          properties: {
            locale: { type: 'string', format: 'bcp47' },
          },
        };
      });

      [
        'en-US',
        'nl-NL',
        'xx-XX',
      ].forEach((locale) => {
        it(`it matches ${locale}`, () => {
          expect({ locale }).toMatchSchemaWithFormatsUnderTest(schema);
        });
      });

      [
        'en-us',
        'EN-US',
        'en_US',
        'enus',
        '123',
      ].forEach((locale) => {
        it(`it does not match ${locale}`, () => {
          expect(() => expect({ locale }).toMatchSchemaWithFormatsUnderTest(schema))
            .toThrowErrorMatchingSnapshot();
        });
      });
    });
  });

  describe('custom keywords', () => {
    it('typeof', () => {
      expect('test').toMatchSchemaWithOptionsUnderTest({
        typeof: 'string',
      });

      // Check error is thrown by custom keyword
      expect(() => expect(false).toMatchSchemaWithOptionsUnderTest({
        typeof: 'string',
      }))
        .toThrowErrorMatchingSnapshot();
    });

    it('instanceof', () => {
      expect([]).toMatchSchemaWithOptionsUnderTest({
        instanceof: 'Array',
      });

      // Check error is thrown by custom keyword
      expect(() => expect(false).toMatchSchemaWithOptionsUnderTest({
        instanceof: 'Array',
      }))
        .toThrowErrorMatchingSnapshot();
    });
  });

  describe('output verbose errors', () => {
    it('should output an error with only the received input printed', () => {
      // Null
      expect(() => expect(null)
        .not.toMatchSchemaWithOptionsUnderTest({
          type: 'null',
        }))
        .toThrowErrorMatchingSnapshot();

      // Boolean
      expect(() => expect(true)
        .not.toMatchSchemaWithOptionsUnderTest({
          type: 'boolean',
        }))
        .toThrowErrorMatchingSnapshot();

      // Number
      expect(() => expect(1)
        .not.toMatchSchemaWithOptionsUnderTest({
          type: 'number',
        }))
        .toThrowErrorMatchingSnapshot();

      // String
      expect(() => expect('this is valid but expect().not.toMatchSchema has been used')
        .not.toMatchSchemaWithOptionsUnderTest({
          type: 'string',
        }))
        .toThrowErrorMatchingSnapshot();

      // Object
      expect(() => expect({})
        .not.toMatchSchemaWithOptionsUnderTest({
          type: 'object',
        }))
        .toThrowErrorMatchingSnapshot();

      // Array
      expect(() => expect(['this is valid but expect().not.toMatchSchema has been used'])
        .not.toMatchSchemaWithOptionsUnderTest({
          minItems: 1,
        }))
        .toThrowErrorMatchingSnapshot();
    });

    it('should display schema $id in the schema path', () => {
      expect(() => expect({
        test: '123',
      }).toMatchSchemaWithOptionsUnderTest({
        $id: 'testSchema',
        allOf: [{
          type: 'object',
          properties: {
            test: {
              type: 'number',
            },
          },
        }],
      }))
        .toThrowErrorMatchingSnapshot();
    });

    it('should output error with details printed per errored property', () => {
      expect(() => expect({
        testType: false,
        testNotEmpty: '',
        testEnum: false,
        testConst: false,
        testFormat: 'test',
        testPattern: 'test',
        testItems: [false, 1],
        testNumber: 1,
        testIf: true,
        testElse: undefined,
        testThen: null,
      }).toMatchSchemaWithOptionsUnderTest({
        $schema: 'http://json-schema.org/draft-07/schema#',
        allOf: [
          {
            type: 'object',
            properties: {
              testType: {
                type: 'null',
              },
              testNotEmpty: {
                type: 'string',
                minLength: 1,
              },
              testEnum: {
                enum: [1, 2],
              },
              testConst: {
                const: true,
              },
              testFormat: {
                format: 'email',
              },
              testPattern: {
                pattern: '[0-9]+',
              },
              testItems: {
                items: {
                  const: true,
                },
                minItems: 5,
                maxItems: 1,
                uniqueItems: true,
                contains: {
                  const: true,
                },
              },
              testNumber: {
                multipleOf: 5,
                minimum: 999,
                maximum: 0,
                exclusiveMinimum: 999,
                exclusiveMaximum: 1,
              },
              testRequired: {
                type: 'string',
              },
            },
            propertyNames: {
              pattern: '^false',
            },
            minProperties: 999,
            maxProperties: 1,
            required: ['testRequired'],
          },
          {
            if: {
              type: 'object',
              properties: {
                testIf: {
                  const: true,
                },
              },
            },
            then: {
              type: 'object',
              properties: {
                testThen: {
                  const: true,
                },
              },
            },
            else: {
              type: 'object',
              properties: {
                testElse: {
                  const: true,
                },
              },
            },
          },
          {
            not: {
              type: 'object',
            },
          },
          {
            if: {
              type: 'object',
              properties: {
                testIf: {
                  const: false,
                },
              },
            },
            then: {
              type: 'object',
              properties: {
                testThen: {
                  const: true,
                },
              },
            },
            else: {
              type: 'object',
              properties: {
                testElse: {
                  const: true,
                },
              },
            },
          },
        ],
      })).toThrowErrorMatchingSnapshot();
    });

    it('ensure verbose readme example is correct', () => {
      const testSchema = {
        $id: 'testVerboseReadmeSchema',
        type: 'object',
        properties: {
          name: {
            type: 'string',
          },
          dob: {
            type: 'string',
            format: 'date',
          },
        },
      };

      const invalidData = {
        name: null,
        dob: '02-29-2000',
      };

      expect(() => {
        expect(invalidData).toMatchSchemaWithOptionsUnderTest(testSchema);
      }).toThrowErrorMatchingInlineSnapshot(`
"expect(received).toMatchSchema(schema)

received
  .name should be string
    Received: <null>
    Path:     testVerboseReadmeSchema#/properties/name/type
  .dob should match format \\"date\\"
    Received: <string> 02-29-2000
    Path:     testVerboseReadmeSchema#/properties/dob/format
"
`);
    });
  });
});
