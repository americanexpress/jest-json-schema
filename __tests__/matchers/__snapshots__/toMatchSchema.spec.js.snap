// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toMatchSchema custom formats bcp47 it does not match 123 1`] = `
"expect(received).toMatchSchema(schema)

received
  .locale should match format \\"bcp47\\"
"
`;

exports[`toMatchSchema custom formats bcp47 it does not match EN-US 1`] = `
"expect(received).toMatchSchema(schema)

received
  .locale should match format \\"bcp47\\"
"
`;

exports[`toMatchSchema custom formats bcp47 it does not match en_US 1`] = `
"expect(received).toMatchSchema(schema)

received
  .locale should match format \\"bcp47\\"
"
`;

exports[`toMatchSchema custom formats bcp47 it does not match en-us 1`] = `
"expect(received).toMatchSchema(schema)

received
  .locale should match format \\"bcp47\\"
"
`;

exports[`toMatchSchema custom formats bcp47 it does not match enus 1`] = `
"expect(received).toMatchSchema(schema)

received
  .locale should match format \\"bcp47\\"
"
`;

exports[`toMatchSchema custom keywords instanceof 1`] = `
"expect(received).toMatchSchema(schema)

received
  should pass \\"instanceof\\" keyword validation
    Expected: <string> Array
    Path:     #/instanceof
"
`;

exports[`toMatchSchema custom keywords typeof 1`] = `
"expect(received).toMatchSchema(schema)

received
  should pass \\"typeof\\" keyword validation
    Expected: <string> string
    Path:     #/typeof
"
`;

exports[`toMatchSchema does not crash on circular references 1`] = `
"expect(received).toMatchSchema(schema)

received
  .hello should be string
"
`;

exports[`toMatchSchema fails for matching schema when using .not 1`] = `
"expect(received).not.toMatchSchema(schema)

Expected value not to match schema

"
`;

exports[`toMatchSchema fails for missing required keys 1`] = `
"expect(received).toMatchSchema(schema)

received
  should have required property 'hello'
"
`;

exports[`toMatchSchema fails for wrong type 1`] = `
"expect(received).toMatchSchema(schema)

received
  .hello should be string
"
`;

exports[`toMatchSchema fails when additional properties are found but forbidden 1`] = `
"expect(received).toMatchSchema(schema)

received
  should NOT have additional properties, but found 'another'
"
`;

exports[`toMatchSchema fails when pattern does not match 1`] = `
"expect(received).toMatchSchema(schema)

received
  .hello should match pattern \\"[a-z]+\\"
"
`;

exports[`toMatchSchema includes the description in the error when provided 1`] = `
"expect(received).toMatchSchema(schema)

en-US language pack
  .hello should be string
"
`;

exports[`toMatchSchema output verbose errors should display schema $id in the schema path 1`] = `
"expect(received).toMatchSchema(schema)

received
  .test should be number
    Received: <string> 123
    Path:     testSchema#/allOf/0/properties/test/type
"
`;

exports[`toMatchSchema output verbose errors should output an error with only the received input printed 1`] = `
"expect(received).not.toMatchSchema(schema)

Expected value not to match schema

received
<null>
"
`;

exports[`toMatchSchema output verbose errors should output an error with only the received input printed 2`] = `
"expect(received).not.toMatchSchema(schema)

Expected value not to match schema

received
<boolean> true
"
`;

exports[`toMatchSchema output verbose errors should output an error with only the received input printed 3`] = `
"expect(received).not.toMatchSchema(schema)

Expected value not to match schema

received
<number> 1
"
`;

exports[`toMatchSchema output verbose errors should output an error with only the received input printed 4`] = `
"expect(received).not.toMatchSchema(schema)

Expected value not to match schema

received
<string> this is valid but expect().not.toMatchSchema has been used
"
`;

exports[`toMatchSchema output verbose errors should output an error with only the received input printed 5`] = `
"expect(received).not.toMatchSchema(schema)

Expected value not to match schema

received
<object> {}
"
`;

exports[`toMatchSchema output verbose errors should output an error with only the received input printed 6`] = `
"expect(received).not.toMatchSchema(schema)

Expected value not to match schema

received
<array> [\\"this is valid but expect().not.toMatchSchema has been used\\"]
"
`;

exports[`toMatchSchema output verbose errors should output error with details printed per errored property 1`] = `
"expect(received).toMatchSchema(schema)

received
  should NOT have more than 1 properties
    Path:     #/allOf/0/maxProperties
  should NOT have fewer than 999 properties
    Path:     #/allOf/0/minProperties
  should match pattern \\"^false\\"
    Path:     #/allOf/0/propertyNames/pattern
  property name 'testType' is invalid
    Expected: {\\"pattern\\":\\"^false\\"}
    Path:     #/allOf/0/propertyNames
  should match pattern \\"^false\\"
    Path:     #/allOf/0/propertyNames/pattern
  property name 'testNotEmpty' is invalid
    Expected: {\\"pattern\\":\\"^false\\"}
    Path:     #/allOf/0/propertyNames
  should match pattern \\"^false\\"
    Path:     #/allOf/0/propertyNames/pattern
  property name 'testEnum' is invalid
    Expected: {\\"pattern\\":\\"^false\\"}
    Path:     #/allOf/0/propertyNames
  should match pattern \\"^false\\"
    Path:     #/allOf/0/propertyNames/pattern
  property name 'testConst' is invalid
    Expected: {\\"pattern\\":\\"^false\\"}
    Path:     #/allOf/0/propertyNames
  should match pattern \\"^false\\"
    Path:     #/allOf/0/propertyNames/pattern
  property name 'testFormat' is invalid
    Expected: {\\"pattern\\":\\"^false\\"}
    Path:     #/allOf/0/propertyNames
  should match pattern \\"^false\\"
    Path:     #/allOf/0/propertyNames/pattern
  property name 'testPattern' is invalid
    Expected: {\\"pattern\\":\\"^false\\"}
    Path:     #/allOf/0/propertyNames
  should match pattern \\"^false\\"
    Path:     #/allOf/0/propertyNames/pattern
  property name 'testItems' is invalid
    Expected: {\\"pattern\\":\\"^false\\"}
    Path:     #/allOf/0/propertyNames
  should match pattern \\"^false\\"
    Path:     #/allOf/0/propertyNames/pattern
  property name 'testNumber' is invalid
    Expected: {\\"pattern\\":\\"^false\\"}
    Path:     #/allOf/0/propertyNames
  should match pattern \\"^false\\"
    Path:     #/allOf/0/propertyNames/pattern
  property name 'testIf' is invalid
    Expected: {\\"pattern\\":\\"^false\\"}
    Path:     #/allOf/0/propertyNames
  should match pattern \\"^false\\"
    Path:     #/allOf/0/propertyNames/pattern
  property name 'testElse' is invalid
    Expected: {\\"pattern\\":\\"^false\\"}
    Path:     #/allOf/0/propertyNames
  should match pattern \\"^false\\"
    Path:     #/allOf/0/propertyNames/pattern
  property name 'testThen' is invalid
    Expected: {\\"pattern\\":\\"^false\\"}
    Path:     #/allOf/0/propertyNames
  .testType should be null
    Received: <boolean> false
    Path:     #/allOf/0/properties/testType/type
  .testNotEmpty should NOT be shorter than 1 characters
    Received: <empty string>
    Path:     #/allOf/0/properties/testNotEmpty/minLength
  .testEnum should be equal to one of the allowed values
    Expected: [1,2]
    Received: <boolean> false
    Path:     #/allOf/0/properties/testEnum/enum
  .testConst should be equal to constant
    Expected: <boolean> true
    Received: <boolean> false
    Path:     #/allOf/0/properties/testConst/const
  .testFormat should match format \\"email\\"
    Received: <string> test
    Path:     #/allOf/0/properties/testFormat/format
  .testPattern should match pattern \\"[0-9]+\\"
    Received: <string> test
    Path:     #/allOf/0/properties/testPattern/pattern
  .testItems should NOT have more than 1 items
    Received: <array> [false,1]
    Path:     #/allOf/0/properties/testItems/maxItems
  .testItems should NOT have fewer than 5 items
    Received: <array> [false,1]
    Path:     #/allOf/0/properties/testItems/minItems
  .testItems[0] should be equal to constant
    Expected: <boolean> true
    Received: <boolean> false
    Path:     #/allOf/0/properties/testItems/items/const
  .testItems[1] should be equal to constant
    Expected: <boolean> true
    Received: <number> 1
    Path:     #/allOf/0/properties/testItems/items/const
  .testItems[0] should be equal to constant
    Expected: <boolean> true
    Received: <boolean> false
    Path:     #/allOf/0/properties/testItems/contains/const
  .testItems[1] should be equal to constant
    Expected: <boolean> true
    Received: <number> 1
    Path:     #/allOf/0/properties/testItems/contains/const
  .testItems should contain a valid item
    Expected: {\\"const\\":true}
    Received: <array> [false,1]
    Path:     #/allOf/0/properties/testItems/contains
  .testNumber should be <= 0
    Received: <number> 1
    Path:     #/allOf/0/properties/testNumber/maximum
  .testNumber should be > 999
    Received: <number> 1
    Path:     #/allOf/0/properties/testNumber/exclusiveMinimum
  .testNumber should be multiple of 5
    Received: <number> 1
    Path:     #/allOf/0/properties/testNumber/multipleOf
  should have required property 'testRequired'
    Path:     #/allOf/0/required
  .testThen should be equal to constant
    Expected: <boolean> true
    Received: <null>
    Path:     #/allOf/1/then/properties/testThen/const
  should match \\"then\\" schema
    Expected: {\\"type\\":\\"object\\",\\"properties\\":{\\"testThen\\":{\\"const\\":true}}}
    Received: {\\"testType\\":false,\\"testNotEmpty\\":\\"\\",\\"testEnum\\":false,\\"testConst\\":false,\\"testFormat\\":\\"test\\",\\"testPattern\\":\\"test\\",\\"testItems\\":[false,1],\\"testNumber\\":1,\\"testIf\\":true,\\"testThen\\":null}
    Path:     #/allOf/1/if
  should NOT be valid
    Expected: {\\"type\\":\\"object\\"}
    Received: {\\"testType\\":false,\\"testNotEmpty\\":\\"\\",\\"testEnum\\":false,\\"testConst\\":false,\\"testFormat\\":\\"test\\",\\"testPattern\\":\\"test\\",\\"testItems\\":[false,1],\\"testNumber\\":1,\\"testIf\\":true,\\"testThen\\":null}
    Path:     #/allOf/2/not
"
`;
